Packet++/src/DoIpLayer.cpp

```diff
@@ -0,0 +1,214 @@
+		}
+		else
+		{
+			return "Unknown Protocol Version";
+		}
+	}
+
+	void DoIpLayer::setProtocolVersion(DoIpProtocolVersion version)
+	{
+		getDoIpHeader()->protocolVersion = static_cast<uint8_t>(version);
+	}
+
+	uint8_t DoIpLayer::getInvertProtocolVersion() const
+	{
+		return getDoIpHeader()->invertProtocolVersion;
+	}
+
+	void DoIpLayer::setInvertProtocolVersion(uint8_t iVersion)
+	{
+		getDoIpHeader()->invertProtocolVersion = iVersion;
+	}
+
+	DoIpPayloadTypes DoIpLayer::getPayloadType() const
+	{
+		return static_cast<DoIpPayloadTypes>(be16toh(getDoIpHeader()->payloadType));
+	}
+
+	void DoIpLayer::setPayloadType(DoIpPayloadTypes type)
+	{
+		getDoIpHeader()->payloadType = htobe16((uint16_t)type);
+	}
+
+	std::string DoIpLayer::getPayloadTypeAsStr() const
+	{
+		auto it = DoIpEnumToStringPayloadType.find(getPayloadType());
+		if (it != DoIpEnumToStringPayloadType.end())
+		{
+			return it->second;
+		}
+		else
+		{
+			return "Unknown Payload type";
+		}
+	}
+
+	uint32_t DoIpLayer::getPayloadLength() const
+	{
+		return htobe32(getDoIpHeader()->payloadLength);
+	}
+
+	void DoIpLayer::setPayloadength(uint32_t Payloadength)
+	{
+		getDoIpHeader()->payloadLength = be32toh(Payloadength);
+	}
+
+	bool DoIpLayer::resolveProtocolVersion() const
+	{
+		DoIpProtocolVersion version = getProtocolVersion();
+		uint8_t inVersion = getInvertProtocolVersion();
+		DoIpPayloadTypes type = getPayloadType();
+
+		// Idea is token from wireshark
+		if (!(version == DoIpProtocolVersion::version01Iso2010 || version == DoIpProtocolVersion::version02Iso2012 ||
+		      version == DoIpProtocolVersion::version03Iso2019 ||
+		      version == DoIpProtocolVersion::version04Iso2019_AMD1 ||
+		      (version == DoIpProtocolVersion::defaultVersion &&
+		       (type >= DoIpPayloadTypes::VEHICLE_IDENTIFICATION_REQUEST &&
+		        type <= DoIpPayloadTypes::VEHICLE_IDENTIFICATION_REQUEST_WITH_VIN))))
+		{
+			PCPP_LOG_ERROR("Invalid/unsupported DoIP version!");
+			return false;
+		}
+		if ((uint8_t)(version) != (uint8_t) ~(inVersion))
+		{
+			PCPP_LOG_ERROR("Version and invert version are not synchronised !");
+			return false;
+		}
+		return true;
+	}
+
+	bool DoIpLayer::resolvePayloadLength() const
+	{
+		uint32_t length = getPayloadLength();
+
+		if (m_DataLen < sizeof(doiphdr))
+		{
+			PCPP_LOG_ERROR("Payload length is smaller than the minimum header size");
+			return false;
+		}
+
+		if (length != (m_DataLen - sizeof(doiphdr)))
+		{
+			PCPP_LOG_ERROR("Payload length does not match expected size");
+			return false;
+		}
+
+		return true;
+	}
+	bool DoIpLayer::resolveLayer() const
+	{
+		// Validate the protocol version and payload length
+		if (!resolveProtocolVersion() || !resolvePayloadLength())
+		{
+			PCPP_LOG_ERROR("Failed to Parse DoIP layer");
+			return false;
+		}
+		return true;
+	}
+
+	std::string DoIpLayer::toString() const
+	{
+		std::stringstream os;
+		if (!resolveLayer())
+		{
+			os << "Malformed doip Packet";
+			return os.str();
+		}
```

Comment:
As this specific branch returns a static string the stream can be skipped here.

```suggestion
		if (!resolveLayer())
		{
			return "Malformed doip Packet";
		}
		std::stringstream os;
```